### PR TITLE
Col refactor

### PR DESCRIPTION
**Pull request by kerolasa**
Welcome back from vacation @karelzak 

While back when d8bfcb4cc5d65239e7c21b1067b82a0b24ae386b was in review phase I mentioned col(1) is in need for whole bunch of attention. This pull request contains the things that I thought make sense when bringing the command from past to present. The patches start with tests that work when they are added, and do not need changes at any step on the way. I hope that gives more confidence these changes are probably ok.

### CODE REVIEW

**Comment by evverx** [comment]
@kerolasa I wonder if it would be possible to run `$TS_CMD_COL` with `ts_run`? It's the only place where ASAN_OPTIONS and UBSAN_OPTIONS are set correctly: https://github.com/karelzak/util-linux/issues/1072.

**Comment by kerolasa** [comment]
> @kerolasa I wonder if it would be possible to run `$TS_CMD_COL` with `ts_run`? It's the only place where ASAN_OPTIONS and UBSAN_OPTIONS are set correctly: #1072.

I can certainly try. Lets see if 2f733b45b52f2fe03197113ae18ec177c2dd3d30 works better.

**Comment by kerolasa** [comment]
I should have taken a look what failed. The cf7825e4e50a994b8f55986aee80432d6495c324 should fix the LeakSanitizer in cost of being free() before exit fix that is usually considered quite pointless activity.

**Comment by evverx** [comment]
I think by analogy with https://github.com/karelzak/util-linux/issues/1077#issuecomment-651050099 another option would be to pass `detect_leaks=0` when the col tests are run. Though all those memory leaks are still on Coverity Scan: https://scan.coverity.com/projects/karelzak-util-linux. I'd say going through them and figuring out what is intentional isn't exactly meaningful activity either :-)

**Comment by evverx** [comment]
Speaking of Coverity @kerolasa I'm not sure if you have access to it. I've just invited you to the "util-linux" organization there. I hope it's all right.

**Comment by evverx** [comment]
FWIW to judge from https://travis-ci.org/github/karelzak/util-linux/jobs/716144038 it seems on macOS `col/io` is failing with
```
diff-{{{

--- /Users/travis/build/karelzak/util-linux/tests/expected/col/io	2020-08-08 16:10:09.000000000 +0000

+++ /Users/travis/build/karelzak/util-linux/tests/output/col/io	2020-08-08 16:27:10.000000000 +0000

@@ -37,7 +37,6 @@

 half line

 9
1

 9exit sane

-col: failed on line 1: Invalid or incomplete multibyte or wide character

 flushing

 1

 2

}}}-diff



 FAILED (col/io)
```

**Comment by kerolasa** [comment]
> Speaking of Coverity @kerolasa I'm not sure if you have access to it. I've just invited you to the "util-linux" organization there. I hope it's all right.

Thank you. Looks like there are plenty of small fixes that could be done. Perhaps I should cease the opportunity and make this release cycle to be mostly about covery fixes.

I will try to look into latest failures. Maybe the easiest is to compile the col(1) on my laptop with leaksanitizer enabled and make the feedback loop a lot shorter.

**Comment by evverx** [comment]
@kerolasa apart from Coverity `util-linux` was integrated into OSS-Fuzz recently (I'll try to start covering more code once https://github.com/karelzak/util-linux/pull/1068 is merged). If you're interested in receiving OSS-Fuzz bug reports let me know so that I can add your email address to https://github.com/google/oss-fuzz/blob/master/projects/util-linux/project.yaml. I should add that non-gmail addresses can only be used to get email notifications. To log in to their bug tracker and to oss-fuzz.com gmail addresses should be used unfortunately.

**Comment by kerolasa** [comment]
@evverx I have no idea what to expect with OSS-Fuzz but I am sure the reports themselves will educate me how to use them. My addresses are kerolasa(at)iki.fi and kerolasa(at)gmail.com. Thank you Evgeny, really nice to see another active contributor working to get util-linux better.

**Comment by evverx** [comment]
> I have no idea what to expect with OSS-Fuzz

OSS-Fuzz has found two issues so far: https://bugs.chromium.org/p/oss-fuzz/issues/list?q=label%3AProj-util-linux&can=1. https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=23861&q=label%3AProj-util-linux&can=1 seems to be a bug in the fuzz target though. Looks like I forgot to limit the number of bytes it's supposed to handle. I'll fix it next week.

> My addresses are ...

I've just opened https://github.com/google/oss-fuzz/pull/4287.

**Comment by kerolasa** [comment]
Hopefully last fix. The col io test "exit sane" is now discarding error message about invalid multibyte because linux and macos does not share the same error message string.

**Comment by kerolasa** [comment]
Really weird, why MacOS is adding newline?

https://travis-ci.org/github/karelzak/util-linux/jobs/716341634#L3804

**Comment by kerolasa** [comment]
The part of test that upset macos is removed.

File: text-utils/col.c [hunk 1]
```diff
@@ -154,9 +150,10 @@ static void __attribute__((__noreturn__)) usage(void)
 	exit(EXIT_SUCCESS);
 }
 
-static void __attribute__((__noreturn__)) wrerr(void)
+static inline void col_putchar(wchar_t ch)
 {
-	errx(EXIT_FAILURE, _("write error"));
+	if (putwchar(ch) == WEOF)
+		errx(EXIT_FAILURE, _("write error"));
```
Comment (karelzak) [inline text-utils/col.c]: Would be better to use _("write failed") as we have in other tools? Would be possible to use err() to get  details about the error?

File: text-utils/col.c [hunk 2]
```diff
@@ -167,21 +167,21 @@ static inline void col_putchar(wchar_t ch)
...
+	if (ctl->nblank_lines & 1) {
 		if (ctl->fine)
 			half = 1;
 		else
-			nb++;
+			ctl->nblank_lines++;
 	}
-	nb /= 2;
-	for (i = nb; --i >= 0;)
+	ctl->nblank_lines /= 2;
```
Comment (karelzak) [inline text-utils/col.c:207]: Is it correct? It seems you modify global ctl->nblank_lines, but the original code uses a local variable and it does not refresh the original global setting -- nblank_lines is unmodified by this function in original code.

File: text-utils/col.c [hunk 3]
```diff
@@ -104,19 +104,20 @@ struct line_str {
 	CHAR	*l_line;		/* characters on the line */
 	LINE	*l_prev;		/* previous line */
 	LINE	*l_next;		/* next line */
```
Comment (karelzak) [inline text-utils/col.c:110]: It would be nice by a separate patch to remove all the typedefs and LINE and CHAR, and use `struct col_line` and `struct col_char`. The typedef makes sense for simple opaque types (like some numbers etc.), otherwise typedef is evil.

File: text-utils/col.c [hunk 4]
```diff
@@ -173,7 +176,7 @@ static void __attribute__((__noreturn__)) usage(void)
 static inline void col_putchar(wchar_t ch)
 {
 	if (putwchar(ch) == WEOF)
-		errx(EXIT_FAILURE, _("write error"));
+		err(EXIT_FAILURE, _("write error"));
```
Comment (karelzak) [inline text-utils/col.c:187]: Ah, yes, err() -- thanks.

File: text-utils/col.c [hunk 5]
```diff
@@ -651,5 +682,6 @@ int main(int argc, char **argv)
 		/* missing a \n on the last line? */
 		ctl.nblank_lines = 2;
 	flush_blanks(&ctl);
+	free_line_allocations(ctl.alloc_root);
```
Comment (karelzak) [inline text-utils/col.c:685]: Maybe we can use some #ifdef to call free_line_allocations() only to make LeakSanitizer happy :-)

File: text-utils/col.c (hunk 1 above)
Comment (kerolasa) [reply text-utils/col.c]: Text is fixed in 3375fd9ccf, and I checked  the err() is used not the errx().

File: text-utils/col.c (hunk 2 above)
Comment (kerolasa) [reply text-utils/col.c]: Good point. I removed the commit that had ill thought variable removal.

File: text-utils/col.c (hunk 3 above)
Comment (kerolasa) [reply text-utils/col.c]: Good idea, typedefs are removed in 2446db107019c6cdfc99cdfea245870ea7b19b78.

File: text-utils/col.c (hunk 5 above)
Comment (kerolasa) [reply text-utils/col.c:698]: Lets see if 3252914310e7a8d76f1a3fab7371a2ff278b2419 works as expected. I added `#if defined(__SANITIZE_ADDRESS__)` around code that is performing these pointless at exit free's.

**Comment by karelzak** [comment]
Thanks! I did tiny changes to the code too.
